### PR TITLE
TRUNK-6189: Provider Service class should provide ability to getProviderAttributeTypeByName

### PR DESCRIPTION
**Pull request by mherman22**
## Description of what I changed

The implemented behavior in this commit is that; - when you create an attribute with a name "test", you cannot create another one with the same name not until you retire the first one. When you retire "test", you create a new one with the name "test" since the existing one has been retired.

This implementation i believed comes out of the discussion we had in the draft pull request at https://github.com/openmrs/openmrs-core/pull/4374 that i will close to pave way for this one.

Below is the implementation in video.

https://github.com/openmrs/openmrs-core/assets/47120265/a7615434-c936-4a4d-8f0a-6c9f27cbbfc1

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6189

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/ProviderService.java [hunk 1]
```diff
@@ -233,6 +234,16 @@ public List<Provider> getProviders(String query, Integer start, Integer length,
...
 	
+	/**
+	 * Get a provider attribute type by it's name
+	 *
+	 * @param name the name of the provider attribute type
+	 * @return the provider attribute type for the given name
+	 * <strong>Should</strong> get the provider attribute type by it's name
+	 */
+	@Authorized({PrivilegeConstants.GET_PROVIDER_ATTRIBUTE_TYPES})
+	ProviderAttributeType getProviderAttributeTypeByName(String name);
```
[the remark this answers is not in the document]
Comment (mherman22) [reply api/src/main/java/org/openmrs/api/ProviderService.java]: done!

File: api/src/main/java/org/openmrs/api/ProviderService.java [hunk 2]
```diff
@@ -20,6 +20,7 @@
 import org.openmrs.annotation.Authorized;
 import org.openmrs.annotation.Handler;
 import org.openmrs.util.PrivilegeConstants;
+import org.springframework.transaction.annotation.Transactional;
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/ProviderService.java:23]: Where are we using this?

Comment (mherman22) [reply api/src/main/java/org/openmrs/api/ProviderService.java:23]: Will remove it in next commit, thanks

File: api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java [hunk 3]
```diff
@@ -40,21 +57,51 @@ public void validate_shouldPassValidationIfFieldLengthsAreCorrect() {
 		assertFalse(errors.hasErrors());
 	}
 	
+	/**
+	 * @see ProviderAttributeTypeValidator#validate(Object,Errors)
+	 */
+	@Test
+	public void validate_shouldFailValidationIfDatatypeClassnameIsEmpty() {
+		ProviderAttributeType type = new ProviderAttributeType();
+		type.setName("name");
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java:66]: How is this related to the ticket?

File: api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java [hunk 4]
```diff
@@ -23,6 +26,20 @@
  */
 public class ProviderAttributeTypeValidatorTest extends BaseContextSensitiveTest {
 	
+	protected static final String PROVIDER_ATTRIBUTE_DATA_XML = "org/openmrs/api/include/ProviderServiceTest-providerAttributes.xml";
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java:29]: Is there a reason why this is not private?

File: api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java (hunk 3 above)
Comment (mherman22) [reply api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java:66]: Thought I'd write the tests for stuff that wasn't covered for the validate method.

File: api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java (hunk 4 above)
Comment (mherman22) [reply api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java:29]: Oversight, thanks. Will edit that

File: api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java (hunk 3 above)
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/validator/ProviderAttributeTypeValidatorTest.java:66]: No big deal though. You can leave it in. :)
